MoreLinq/SelectAsync.cs

```diff
@@ -0,0 +1,439 @@
+
+        /// <summary>
+        /// Returns a new asynchronous projection operation with the given
+        /// Boolean indicating whether or not the projections should be
+        /// returned in the order of the projection source.
+        /// </summary>
+
+        public static ISelectAsyncEnumerable<T> PreserveOrder<T>(this ISelectAsyncEnumerable<T> source, bool value)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            return source.WithOptions(source.Options.WithPreserveOrder(value));
+        }
+
+        /// <summary>
+        /// Asynchronously projects each element of a sequence to its new form.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// This method uses deferred execution semantics. The results are
+        /// yielded as each asynchronous projection completes and, by default,
+        /// not guaranteed to be based on the source sequence order. If order
+        /// is important, compose further with
+        /// <see cref="PreserveOrder{T}(ISelectAsyncEnumerable{T})"/>.</para>
+        /// <para>
+        /// This method starts a new task where the asynchronous projections
+        /// are started and awaited.</para>
+        /// <para>
+        /// The <paramref name="selector"/> function should be designed to be
+        /// thread-agnostic.</para>
+        /// </remarks>
+
+        public static ISelectAsyncEnumerable<TResult> SelectAsync<T, TResult>(
+            this IEnumerable<T> source, Func<T, Task<TResult>> selector)
+        {
+            if (selector == null) throw new ArgumentNullException(nameof(selector));
+            return source.SelectAsync((e, _) => selector(e));
+        }
+
+        /// <summary>
+        /// Asynchronously projects each element of a sequence to its new form.
+        /// The projection function receives a <see cref="CancellationToken"/>
+        /// as an additional argument that can be used to abort any asynchronous
+        /// operations in flight.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// This method uses deferred execution semantics. The results are
+        /// yielded as each asynchronous projection completes and, by default,
+        /// not guaranteed to be based on the source sequence order. If order
+        /// is important, compose further using
+        /// <see cref="PreserveOrder{T}(MoreLinq.ISelectAsyncEnumerable{T})"/>
+        /// and a Boolean value of <c>true</c>.</para>
+        /// <para>
+        /// This method starts a new task where the asynchronous projections
+        /// are started and awaited.</para>
+        /// <para>
+        /// The <paramref name="selector"/> function should be designed to be
+        /// thread-agnostic.</para>
+        /// </remarks>
+
+        public static ISelectAsyncEnumerable<TResult> SelectAsync<T, TResult>(
+            this IEnumerable<T> source, Func<T, CancellationToken, Task<TResult>> selector)
```

Comment:
We have to be careful about passing “live” items into a query. The query is a declaration of a computation that is then evaluated, possible at some point in the future. The `CancellationToken` can introduce side-effects that breaks subsequent evaluations.

`SelectAsync`, which I am thinking about renaming, is very misleading as a name. It's not async because it returns a sequence. It's the projection that's async. Someone who is async-heavy won't be making use of it because it can cause the thread calling `SelectAsync` to block. `SelectAsync` is good in synchronous code that requires I/O or asynchronous operation for the projections.